### PR TITLE
Fix bug in devicemapper event handling

### PR DESCRIPTION
**Pull request by jbaublitz**
Blocked by #2508
Closes stratis-storage/project#280

### CODE REVIEW

File: src/engine/types/mod.rs [hunk 1]
```diff
@@ -353,3 +353,9 @@ impl<'a> From<&'a libudev::Device<'a>> for UdevEngineDevice {
         }
     }
 }
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct DevicemapperInfo {
```
Comment (mulkieran) [inline src/engine/types/mod.rs:358]: I'm having trouble justifying why this summary information is a sufficient proxy, for two reasons:

1. What if the highest event_nr is one that hasn't changed? For example, for some reason the thinpool has the highest event number, but there's a filesystem that has experienced a change, and has had it's event number bumped. I think the pool should still ```event_on``` in that case.
2. ```device_count``` might be the same if a filesystem had been removed and another one added. But in that case, I think an ```event_on``` would still be indicated.

Maybe ```DevicemapperInfo``` can only be accurate if it's just a map of names to event numbers?

**Review by mulkieran — CHANGES_REQUESTED** [review]
Code looks great, but I do have one general question.

File: src/engine/types/mod.rs (hunk 1 above)
Comment (jbaublitz) [reply src/engine/types/mod.rs:360]: Okay, this should be resolved!

File: src/engine/strat_engine/engine.rs [hunk 2]
```diff
@@ -303,16 +301,35 @@ impl Engine for StratEngine {
...
+                .iter()
+                .filter_map(|(dm_name, event_nr)| {
+                    if dev_names.contains(dm_name) {
+                        Some((dm_name.clone(), *event_nr))
+                    } else {
+                        None
+                    }
+                })
+                .collect::<HashMap<_, _>>();
+            let device_count = event_nrs.len();
```
Comment (mulkieran) [inline src/engine/strat_engine/engine.rs:316]: I think ```device_count``` is now redundant.

**Review by mulkieran — CHANGES_REQUESTED** [review]

File: src/engine/strat_engine/engine.rs (hunk 2 above)
Comment (jbaublitz) [reply src/engine/strat_engine/engine.rs]: Resolved!

**Review by mulkieran — CHANGES_REQUESTED** [review]
Looks great! Please close this and open a final PR.

**Pull request closed without merging**